client/web/src/enterprise/batches/detail/BatchChangeDetailsPage.tsx

```
@@ -40,56 +42,45 @@ export interface BatchChangeDetailsPageProps extends BatchChangeDetailsProps {
-    const {
-        namespaceID,
-        batchChangeName,
-        history,
-        location,
-        telemetryService,
-        fetchBatchChangeByNamespace: fetchBatchChangeByNamespace = _fetchBatchChangeByNamespace,
-        deleteBatchChange,
-    } = props
+    const { namespaceID, batchChangeName, history, location, telemetryService, deleteBatchChange } = props
 
     useEffect(() => {
         telemetryService.logViewEvent('BatchChangeDetailsPage')
     }, [telemetryService])
 
-    const createdAfter = useMemo(() => subDays(new Date(), 3).toISOString(), [])
-    const batchChange: BatchChangeFields | null | undefined = useObservable(
-        useMemo(
-            () =>
-                fetchBatchChangeByNamespace(namespaceID, batchChangeName, createdAfter).pipe(
-                    repeatWhen(notifier => notifier.pipe(delay(5000))),
-                    distinctUntilChanged((a, b) => isEqual(a, b))
-                ),
-            [fetchBatchChangeByNamespace, namespaceID, batchChangeName, createdAfter]
-        )
+    // Query bulk operations created after this time.
+    const createdAfter = useMemo(() => subDays(startOfDay(new Date()), 3).toISOString(), [])
+
+    const { data, error, loading } = useQuery<BatchChangeByNamespaceResult, BatchChangeByNamespaceVariables>(
+        BATCH_CHANGE_BY_NAMESPACE,
+        {
+            variables: { namespaceID, batchChange: batchChangeName, createdAfter },
+            fetchPolicy: 'cache-and-network',
+            // TODO: Why do we need to poll this every 5 seconds??
+            // pollInterval: 5000,
+        }
     )
 
-    // Is loading.
-    if (batchChange === undefined) {
+    if (loading && !data) {
         return (
             <div className="text-center">
                 <LoadingSpinner className="icon-inline mx-auto my-4" />
             </div>
         )
     }
-    // Batch change was not found.
-    if (batchChange === null) {
+    if (error || !data || !data.batchChange) {
```

I must admit I don't really like this overwhelming set of different states representable with all the env vars.

---

And there is a bug in here, which would hide actual fetch errors and would always display the 404 page. It should rather throw the error so the ErrorBoundary can take care of it, I think.

---

Ooh I didn't realize we had an error boundary for that, thanks for calling that out! It's definitely a bit less overwhelming now since the case with just `error` will be separate for the error boundary, but I think the `!data || !data.batchChange` bit will just have to be the sacrifice for pulling it out of the cache. 🙁